### DISCUSSION TITLE
Standardization vs Min-Max Normalization for Data Scaling

### DISCUSSION THREAD

**Discussion by m-julian** [cornellius-gp/gpytorch#1766] [General]
Hello,

I am curious as to why standardization is recommended for both inputs and outputs for GP regression over other scaling methods such as min-max normalization. If training outputs are standardized, they have zero mean, so a zero mean GP could be used, after which test predictions can be inverse transformed (using the mean and standard deviation of the training data) to get the "true" predicted values. For inputs, it makes sense to scale the data in some way to remove the units associated with each feature, but both standardization and min-max normalization make the data unitless.

Could you use other methods such as min-max normalization to scale the data instead? For example, could you use min-max normalization for the inputs and standardize the outputs? Would using a scaling method other than standardization impact model performance significantly? I have not found many discussions on scaling data for GPs specifically, the best one I've found is this: https://stats.stackexchange.com/q/178245/317677, but it only covers standardization.

Any discussion on data scaling for GPs would be greatly appreciated, thanks!

**Comment by wjmaddox** [comment]
Yes, min-max normalization for inputs to `[0, 1]^d` and standardization of outputs is probably a good idea for numerical stability. For example, these types of transforms are implemented in BoTorch [here](https://github.com/pytorch/botorch/blob/main/botorch/models/transforms/input.py) for example, where `Normalize(d = ...)` is min-max normalization and `Standardize(m=...)` is standardization.

**Comment by m-julian** [comment]
Thanks for the link, I've implemented these transforms myself because I didn't know they were already implemented in BoTorch.

Is there a reason why only standardization is recommended though? I've only seen standardization done in the GPyTorch docs and in github issues here. Are default hyperparameter values better suited for standardization or is it something else that makes standardization better for GPs (for both inputs and outputs)?

**Reply by wjmaddox** [reply]
By standardizing you mean standardizing the outputs to have zero mean and standard deviation one? Yes, the default hyperparameters are setup to somewhat assume this type of data, also, it helps keep everything more numerically stable.

And, it's sort of a soft recommendation, if you know your data is skewed in some way, say positive, it would often make more sense to take a log transform (and then even standardize).

**Reply by m-julian** [reply]
Yep, I refer to zero mean and standard deviation of one. Last question for now: Are there any downsides of scaling the data?

There are 2 minor ones I can think of, but please let me know if I am overlooking something that is more important:
1. Having to apply scaling to the test inputs and reversing the scaling on the predictions to get the "true" predictions. Not really a downside, but something extra that needs to be done. 
2. If some input features are cyclic and the period is known, the period will change if a data transform is applied. This is what I've been having to do when fitting data with a periodic kernel and scaled data.

**Reply by wjmaddox** [reply]
Yes 1) can be tricky but with 2) there's probably a way to adjust the period or to do an input transform to take that into account to properly upscale / downscale the period as necessary.

The larger "gotcha" so to speak is that your parameters (aka kernel lengthscales) will need to be transformed into the original scale if you want to interpret them in a "physical" manner.

**Reply by wjmaddox** [reply]
Also, one day, we will eventually move the transforms back into gpytorch and deal with 1) automatically, see #1652

**Reply by m-julian** [reply]
Great, thanks, appreciate the help!

**Reply by patel-zeel** [reply]
Hi @wjmaddox and @m-julian,

1\) is straightforward for predictive mean transformation but what do we need to do to inverse transform the predictive covariance matrix or at least predictive variance? Is there a straightforward way to do this as well?

**Reply by wjmaddox** [reply]
Yes, you'll also need to back transform the predictive mean and variance. For standardization you need to just follow variance / stddev properties: $\sigma(c (X + b) ) = |c| \sigma(X)$. An example is here: https://github.com/pytorch/botorch/blob/3de2ccd55b41a1ca341772a6a2eb29ce787ecd16/botorch/models/transforms/outcome.py#L294 .
